src/components/Layout/DataTypes.tsx

```diff
@@ -0,0 +1,283 @@
+              depth={depth}
+              dataType={{ ...Contains, Value: v }}
+            />
+          ))}
+        </>
+      )
+
+    case Type.Object:
+      const { Name, Children } = dataType
+
+      if (Name === 'Cid') {
+        const cid = Children['/'].Value as string
+        return (
+          <DataTypeLine
+            label={`${label} (CID)`}
+            depth={depth}
+            dataType={{ Name, Type: Type.String, Value: cid }}
+          />
+        )
+      }
+
+      return (
+        <DataTypeMapLines label={label} depth={depth} dataTypeMap={Children} />
+      )
+
+    default:
+      throw new Error(`Unexpected DataType: ${JSON.stringify(dataType)}`)
+  }
+}
+
+interface DataTypeLinesProps {
+  label?: string
+  depth?: number
+  dataType: DataType
+}
+
+DataTypeLines.propTypes = {
+  label: PropTypes.string,
+  depth: PropTypes.number,
+  dataType: PropTypes.object.isRequired
+}
+
+/**
+ * DataTypeMapLines
+ * Renders a DataTypeMap on multiple lines
+ */
+
+export const DataTypeMapLines = ({
+  label,
+  depth,
+  dataTypeMap
+}: DataTypeMapLinesProps) => (
+  <>
+    {label && <Line label={label} depth={depth} />}
+    {Object.entries(dataTypeMap).map(([key, dt]) => (
+      <DataTypeLines
+        key={key}
+        label={key}
+        depth={label ? (depth ?? 0) + 1 : depth}
+        dataType={dt}
+      />
+    ))}
+  </>
+)
+
+interface DataTypeMapLinesProps {
+  label?: string
+  depth?: number
+  dataTypeMap: DataTypeMap
+}
+
+DataTypeMapLines.propTypes = {
+  label: PropTypes.string,
+  depth: PropTypes.number,
+  dataTypeMap: PropTypes.object.isRequired
+}
+
+/**
+ * DataTypeLine
+ * Renders the value of a DataType on a single line
+ */
+
+export const DataTypeLine = ({ label, depth, dataType }: DataTypeLineProps) => (
+  <Line label={label} depth={depth}>
+    <DataTypeValue dataType={dataType} />
+  </Line>
+)
+
+interface DataTypeLineProps {
+  label?: string
+  depth?: number
+  dataType: DataType
+}
+
+DataTypeLine.propTypes = {
+  label: PropTypes.string,
+  depth: PropTypes.number,
+  dataType: PropTypes.object.isRequired
+}
+
+/**
+ * DataTypeValue
+ * Renders the value of a DataType using the most suitable component.
+ */
+
+export const DataTypeValue = ({ dataType }: DataTypeValueProps) => {
+  const { Name, Value } = dataType
+
+  switch (dataType.Type) {
+    case Type.Bool:
+      const boolVal = Value as boolean
+      return <>{boolVal.toString()}</>
+
+    case Type.String:
+      const strVal = Value as string
+
+      if (Name === 'Address')
+        return (
+          <AddressLink address={strVal} fetchAddress hideCopyText={false} />
+        )
+
+      if (Name === 'FilecoinNumber')
+        return <>{new FilecoinNumber(strVal, 'attofil').toFil() + ' FIL'}</>
+
+      return <>{Value}</>
+
+    case Type.Number:
+      const numVal = Value as number
+      return <>{numVal}</>
+
+    case Type.Array:
+      const arrVal = Value as Array<BaseValue>
```

Comment:
In our current use, can this ever actually happen? It looks like `DataTypeLines` maps over arrays and renders a `DataTypeLine` for each 

Either way, it seems like we should share similar behavior, and instead of joining array values by a string, we should just recurse through the array and render `DataTypeLines` for each array item like we do elsewhere?

Hard for me to get a good sense since i'm having trouble finding a case where this hits

Reply:
That's true. I was considering a use case where we do want to display an array on a single line (maybe a bitfield), but that's not really relevant right now. I'll remove it and we can tackle it again if we need it